README.md

```diff
@@ -1,168 +1,156 @@
+and then decorate certain `<a>` elements with `class="high-likelihood-prerender"`.
 
-### Opt-out only
+When the browser performs this prerendering, it loads the same-origin document in a same-origin prerendering browsing context. Content there is not allowed to do disruptive, user-visible things, but it has full access to credentials, storage, etc.
 
-Could an opt-out suffice? It would seem to make it easier to increase coverage.
+### Cross-origin news aggregator
 
-In principle, yes. However, this would mean that content may become broken (as described above) and require authors to roll out an opt-out to simply restore the previous level of functionality. This seems problematic, especially since we anticipate a large fraction of the web would be affected.
+Consider a news aggregator website, which contains many links to different origins providing news articles.
 
-### HTTP header exchange only
+Some such news providers might be prepared to be prerendered. To do so, they would opt in with the HTTP response header
 
-An opt-in solely based on HTTP request and response headers is appealing. It allows the fetch to be terminated before processing the response body, and doesn't require HTML parsing or other complicated logic to handle.
+```http
+Supports-Loading-Mode: uncredentialed-prerender
+```
 
-The largest drawback here is one of adoption. In order to make as much content as possible available for uncredentialed prerendering, we would like to make it as easy as possible for authors to mark eligible content. We have heard from developers that many of them find it much easier to deploy changes that only affect content than changes which also require server behavior changes, even relatively straightforward ones. For example, these may be managed by different teams or not be possible at all. One key example here is that GitHub Pages doesn't allow users to set response headers.
+or the HTML `<meta>` element
 
-This strongly indicates an opt-in in the HTML document, such as a `<meta>` tag, attribute of the `<html>` element, or other affordance entirely within the HTML response body.
+```html
+<meta http-equiv="Supports-Loading-Mode"
+      content="uncredentialed-prerender">
+```
 
-### Document policy
+As in the last example, based on its heuristics or informed by prerender triggers, the browser can attempt to prerender these linked-to news articles. Since they are cross-origin, however, the process is more restricted. The initial fetch, as well as any subresource fetches, are performed without credentials. If the response that comes back does not have the opt-in, then the result is discarded.
```

Comment:
Is opt-in needed if there were no credentials when the prerender started?

Reply:
This is still unclear to me. There's talk of "adaptive prerendering" which I have a hard time understanding, and could use some help incorporating into this document.

In particular, a model that tries to prerender without opt-in runs into the following issues:

- Does that break impression tracking?
- What happens if the page tries to write to storage?
  - If it fails, then we're likely to break pages. So an opt-in would be better.
  - If it goes to partitioned storage, then we have to figure out how to join the partitioned storage and the unpartitioned storage after transition to a real top-level browsing context. Which seems basically impossible to do automatically. So an opt-in, plus custom page logic to handle the transition, makes more sense.

Reply:
I'm also unsure how we want to reason about adaptive prerendering long-term, and what we can/cannot do without modification.

@kinu

Reply:
Note, in a world where everyone opts-in we don't need this.

My reasoning is that it shouldn't have a tough conflict as far as it can go without requiring cookies (while the site can still opt-out and drop the request). For actions like prefetching this is even more plausible.

No storage writes should happen, they will be deferred as much as possible, or if sync access happens the prerender can bail out.

Exception is the 1p cookie, they will be written back and merged, and this is possible (and we have code) because we're talking about no-storage cases.

Reply:
Adaptive one is an exploration of what we can do *without* conflicting with what we want to be doing.  Therefore there shouldn't be many that need to be incorporated explicitly for adaptive cases, while, one exception is the first main resource response.  If UA has no cookies it doesn't feel that the necessity to have response opt-in is a bit slim.

Reply:
Sorry, I had to struggle to come up with a possible alternative text (and the reasonable behavior).  I think I'm also okay to drop this eventually, but not yet fully comfortable saying that "we require opt-ins for all responses, otherwise UA does nothing".

What do we think if we tweak this like following:

'... are performed without credentials.  This means that the target site may get uncredentialed requests even when there are credentials stored in UA. If the response that comes back does not have the opt-in in such cases, the result is discarded.'

Reply:
kinu's suggestion works for me.

Reply:
Incorporated; thanks!